### PR TITLE
Changes so lando.py works based on a queue 

### PR DESCRIPTION
**Pull request by johnbradley**
Adding changes so lando.py will listen to a AMPQ for tasks and perform them.
Moved VM creation code out to cloudservice.py
Adding workqueue that contains the AMPQ changes
Added lando_client to put items into the work queue(AMPQ)
Added bootscript to build the script that is run when the VM boots

### CODE REVIEW

**Comment by johnbradley** [comment]
# Big Picture
System for running cwl workflows in openstack cloud.
1. A rabbitmq queue that holds launch and terminate instance messages
2. lando.py server that processes messages from the queue creating and terminating VMs
3. lando_client.py program that adds a new message to the queue containing all data necessary to run the workflows

# Walkthrough
User has a CWL workflow that downloads files, runs processes and saves the results.
I have some samples CWL workflows here: https://github.com/johnbradley/toyworkflow.git

## Start lando.py server
Copy landoconfig.sample to landoconfig.yml and fill in settings.
Launch the server:
```
python lando.py
```
This will hang until it is killed or sent the `shutdown_server` command.

## Create job yaml config file
This file has 4 required properties.
* `git_clone` - url to git repo we will clone
* `cd` - directory inside of the repo we will cd into
* `input` - content of the yaml properties file cwl-runner will be passed
* `run_cwl` - name of the workflow we want to run

Example to run fastqc on a fastq file in a DukeDS repo I put into a file named `fastqc-job.yml`:
```
git_clone: https://github.com/johnbradley/toyworkflow.git
cd: toyworkflow
input: |
  dds_agent_key: <somekey>
  dds_user_key: <somekey>
  projectName: MouseRNA
  input_file: ERR550644_1.fastq
  output_file: ERR550644_1_fastqc.zip
run_cwl: fastqc-workflow.cwl
```

## Send job to a queue with lando_client.py
Create workerconfig.yml based on landoconfig.yml.
This script has many positional arguments that take the form:
```
python lando_client.py <config_filename> <command> <job_yaml_filename> 
```
So to request `fastqc-job.yml` from earlier:
```
python lando_client.py workerconfig.yml start_worker fastqc-job.yml 
```

## Job flow
At this point `lando.py` should receive the message.
A VM will be created in openstack and a startup script will be run.
This VM will already have the lando source code installed at `/lando`.
This startup script will 
1. create a workerconfig.yml file for communicating with the queue.
2. create a cwl parameter yaml file for use with cwl-runner
3. run `/lando/lando_worker.sh`. This script will run the cwl workflow based on arguments passed in then queue a message to terminate the VM.




_This isn't meant to be the final way this all works._

**Comment by dleehr** [comment]
FYI, it's AMQP not AMPQ. <insert joke about minding p's and q's>

File: bootscript.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+
+class BootScript(object):
+    """
+    Creates a bash script that will run a CWL workflow.
+    """
+    def __init__(self, yaml_str, worker_config_yml, server_name):
+        """
+        Fills in content property with script based on passed in settings.
+        :param yaml_str: dict: contains url to clone, working directory, name of workflow to run and input yaml
```
Comment (dleehr) [inline bootscript.py]: yaml_str is a dict?

**Comment by dleehr** [comment]
Can you make the worker image public in OpenStack? I can't see it with my credentials.

**Comment by dleehr** [comment]
How should `workerconfig.yml` and `landoconfig.yml` differ?

**Comment by johnbradley** [comment]
landoconfig.yml contains workerconfig.yml.
lando creates the contents of workerconfig.yml based on landoconfig.yml.
Part of the startup script is to write workerconfig.yml out so the worker can send the delete my vm message(which will go away soon).

**Comment by dleehr** [comment]
My question about workerconfig vs landoconfig was based on this bit:

> Create workerconfig.yml based on landoconfig.yml.
> ...
> So to request `fastqc-job.yml` from earlier:
> ```
> python lando_client.py workerconfig.yml start_worker fastqc-job.yml 
> ```

Is that still accurate? From the recent comment it sounds like I don't need to create a workerconfig - that I would supply a landoconfig instead and lando_client would create and deploy the workerconfig

**Comment by johnbradley** [comment]
landoconfig.yml would work, but it has settings that aren't appropriate for a lando_client.
Right now any user can do anything. Long term we may lock it down so they are bound to particular roles.

We might end up with three different queue users.
1) Lando who listens for run_job, cancel_job, job_status, etc
2) Worker who listens for stage, run_workflow, archive, cancel_workflow
3) webserver who sends run_job, cancel_job

**Comment by dleehr** [comment]
OK

**Pull request merged**